### DISCUSSION TITLE
http POST request returns 400 (bad request)

### DISCUSSION THREAD

**Discussion by PaulskPt** [micropython/micropython#16376] [Using MicroPython]
The goal of my project is to send to a Google Sheets spreadsheet the following data:

a value from a MCP9808 temperature sensor;
a datetime string;
a board ID.
To achieve this the MPY script sends a https POST request to a Google Apps Script.

I already did these kind of things for other types of boards and other programming environments (Arduino and Circuitpython).

Hardware used:

a Pimoroni PicoPlus2 (RP2350) microcontroller;
a Pimroni RM2 (WiFi/BLE) module;
an Adafruit MCP9808 temperature sensor.
The PicoPlus2 is flashed with Micropython (MicroPython feature/psram-and-wifi, pico_plus2_rp2350 v0.0.10 on 2024-11-18; Pimoroni Pico Plus 2 (PSRAM + LTE + WiFi) with RP2350).

The data the PicoPlus2 sends is accepted and handled by the Google Apps Script.

The data is arriving in the spreadsheet.

A Logger info writes to a Logger spreadsheet. The result is: "success", however the response to the post request, that the MPY script receives is: 400 (bad request).

I googled for similar cases but did not found any. I created a discussion post in the discussion page of Micropython.org on Github however no reply sofar.

The Google Apps script has been deployed and set for access to anyone. The apikey is limited for websites; with restrictions to spreadsheet where the data has to go; the Logger spreadsheet; https://script.google.com/home/projects/*; https://script.google.com/macros/*; https://www.googleapis.com/auth/spreadsheets/*

Question: anyone who can tell me what could cause the 400 response?

Below I will copy the most important lines related to the POST. I uploaded images of the spreadsheet and its content and of the hardware used: [images](https://imgur.com/a/OEfy8zL).
Extract of the script, function ```handle_request()```:
```
def handle_request():
    yy = 0
    mo = 1
    dd = 2
    hh = 3
    mi = 4
    ss = 5
    ntp_time = ntptime.time()  # get datetime (UNIX TIME) (EPOCH)

    # Example Unix time (seconds since January 1, 1970)
    #unix_time = 1672531199  # Equivalent to 2023-01-01 00:00:00

    # Convert Unix time to a tuple representing the local time
    local_time = utime.localtime(ntp_time)  # GMT

    # Format the local time as a datetime string
    datetimeStr = "{:04d}-{:02d}-{:02d}T{:02d}:{:02d}:{:02d}Z".format(
        local_time[yy], local_time[mo], local_time[dd], 
        local_time[hh], local_time[mi], local_time[ss]
    )

    print("ntp_time = {}".format(datetimeStr), end='\n')
    #datetimeStr = "2024-12-10T00:31:18Z"
    
    tempc = ck_temp()
    print("tempc = {:f}".format(tempc), end='\n')
    jsonPayload = ujson.dumps({'key':GOOGLE_API_KEY,'data':{'date':datetimeStr,'tempc':tempc,'id':'PicoPlus2'}})
    # Get the length of the JSON payload
    content_length = len(jsonPayload)
    
    headers = {'Content-Type': 'application/json'}
    # Add the Content-Length header
    headers['Content-Length'] = str(content_length)
    url = GOOGLE_APPS_SCRIPT_URL
    
    print("headers = {}".format(headers), end='\n')
    #print("url = \"{:s}\"".format(url), end='\n')
    print("jsonPayload = {}".format(jsonPayload), end='\n')
    print("GOOGLE_APPS_SCRIPT_URL = {}".format(GOOGLE_APPS_SCRIPT_URL), end='\n')
    
    try:
        # r = requests.post(GOOGLE_APPS_SCRIPT_URL,data=jsonPayload,headers=header)
        r = requests.post(url,data=jsonPayload,headers=headers)
        print("Response text = {}".format(r.text), end='\n')
        status_code = r.status_code
        print("request status_code = {}".format(status_code), end='\n')
        if status_code == 400:  # = Bad Request
            cleanup()
            raise RuntimeError
            
    except ValueError as e:
        print(f"error: {e}")
        cleanup()
        raise
		
		
Response text = <!DOCTYPE html>
<html lang=en>
  <meta charset=utf-8>
  <meta name=viewport content="initial-scale=1, minimum-scale=1, width=device-width">
  <title>Error 400 (Bad Request)!!1</title>
  <style>

  </style>
  <a href=//www.google.com/><span id=logo aria-label=Google></span></a>
  <p><b>400.</b> <ins>That’s an error.</ins>
  <p>Your client has issued a malformed or illegal request.  <ins>That’s all we know.</ins>
```
I think the request is not ```illegal``` because it is accepted and handled by Google Apps.
Question: anyone who can explain me where the request could be malformed?

**Comment by PaulskPt** [comment]
Today I added to the "headers"
```
    content_length = len(jsonPayload)
    headers = {'Content-Type': 'application/json'}
    # Add the Content-Length header
    headers['Content-Length'] = str(content_length)
```

See also an [image](https://imgur.com/a/m41xbiG) of the jsonPayload as a json file.

however the error 400 kept coming.

**Comment by andrewleech** [comment]
In my honest opinion, this discussion possibly didn't get reponses earlier because it sounds to me like a "google sheets" question more than a microypthon question. You're getting response text from google that's saying your request is somehow malformed. 
I don't know anything about the sheets post api to know what about your raw data is being detected as invalid.

What `requests` library are you using? You could patch it to `print()` out the raw data of the post request just before it sends it so you can see the exact raw data it's sending so you can compare that data to what the sheets api accepts.

**Comment by PaulskPt** [comment]
I don't know which requests library is being used. The board is flashed with a custom version of micropython by Pimoroni. The version is: "MicroPython feature/psram-and-wifi, pico_plus2_rp2350 v0.0.10 on 2024-11-18; Pimoroni Pico Plus 2 (PSRAM + LTE + WiFi) with RP2350. "  I assume that Pimoroni used the latest version of micropython.

In the Google Apps Script  I use I added logging commands. The Logger.log command below prints the raw data received
by the Google Apps Script:

function doPost(e) {
   Logger.log("e = %s", e);
  // etcetera...
}


The Logger.log command resulted in following data the Google Apps Script is receiving:
```
{
    "parameter":{},
    "contentLength":128, 
    "parameters":{},
    "postData":
    {
        "contents":"{
            \"key\": \"<GOOGLE_API_KEY>\",        // Note: the real key value was here
            \"data\": 
            {
                \"id\": \"PicoPlus2\", 
                \"tempc\": 24.875, 
                \"date\": \"2024-12-15T23:51:14Z\"
            }
        }",
        "length":128,
        "name":"postData",
        "type":"application/json"
    },
    "contextPath":"",
    "queryString":""
}
````

**Comment by andrewleech** [comment]
I don't know what Google app scripts are, though I'm presuming by the time data gets to it it's already gone through whatever layer of Google is returning the 400 to you. 

I think you need to try to send the same request raw post data your device actually sends via some other platform like curl or cpython / requests to see if you also get the 400 error there. The issue might be related to redirects or headers or whatever your google app script is returning?

You might need to get some support from pimoroni to find it how to inspect the post data. 
Or just mip instead a different http request library you could then modify to print diagnostics.

The json payload doesn't really say much about the overall post request or whether the issue is related to the http client library in use or the Google API usage.

**Reply by PaulskPt** [reply]
Thank you Andrew for your answer. I am going to follow your advice.
